### PR TITLE
avm2: Port `flash.ui.Keyboard` to ActionScript

### PR DESCRIPTION
**Pull request by relrelb**

### CODE REVIEW

**Comment by Aaron1011** [comment]
It's a little unfortunate that we now duplicate the `KeyCode` values between the Flash class and our `enum KeyCode`

**Comment by relrelb** [comment]
> It's a little unfortunate that we now duplicate the `KeyCode` values between the Flash class and our `enum KeyCode`

Agree, but I guess we'll have to live with it. I can add a comment explaining this.

**Comment by adrian17** [comment]
We could try to make `build_playerglobal` export constants as Rust constants.  Like:
```rs
mod flash { mod ui { mod keyboard { mod constants { 
    const A: uint = 65;
}}}}```
Not as obvious what to do with string constants though.

File: core/src/avm2/globals/flash/ui/Keyboard.as [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+        public static native function get hasVirtualKeyboard(): Boolean;
+
+        public static native function get numLock(): Boolean;
+
+        public static native function get physicalKeyboardType(): String;
+
+        public static native function isAccessible(): Boolean;
+
+        public static const A: uint = 65;
+
```
[the remark this answers is not in the document]
Comment (relrelb) [reply core/src/avm2/globals/flash/ui/Keyboard.as:14]: It doesn't require, will do.

File: core/src/avm2/globals/flash/ui/Keyboard.as [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+        public static const AUDIO: uint = 0x01000017;
+        public static const SUBTITLE: uint = 0x01000018;
+        public static const DVR: uint = 0x01000019;
+        public static const VOD: uint = 0x0100001A;
+        public static const INPUT: uint = 0x0100001B;
+        public static const SETUP: uint = 0x0100001C;
+        public static const HELP: uint = 0x0100001D;
+        public static const MASTER_SHELL: uint = 0x0100001E;
+        public static const SEARCH: uint = 0x0100001F;
+        public static const PLAY_PAUSE: uint = 0x01000020;
```
Comment (danielhjacobs) [inline core/src/avm2/globals/flash/ui/Keyboard.as:141]: https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#PLAY_PAUSE

"This constant is supported on AIR for TV. On other platforms, its value is undefined."

Not sure it matters though.

Comment (danielhjacobs) [reply core/src/avm2/globals/flash/ui/Keyboard.as:141]: Strangely, the documentation says the same thing for all the following constants, but only PLAY_PAUSE is marked with the AIR-only symbol:

- [AUDIO](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#AUDIO)
- [BACK](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#BACK)
- [BLUE](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#BLUE)
- [CHANNEL_DOWN](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#CHANNEL_DOWN)
- [CHANNEL_UP](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#CHANNEL_UP)
- [DVR](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#DVR)
- [EXIT](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#EXIT)
- [FAST_FORWARD](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#FAST_FORWARD)
- [GREEN](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#GREEN)
- [GUIDE](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#GUIDE)
- [HELP](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#HELP)
- [INFO](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#INFO)
- [INPUT](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#INPUT)
- [LAST](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#LAST)
- [LIVE](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#LIVE)
- [MASTER_SHELL](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#MASTER_SHELL)
- [MENU](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#MENU)
- [NEXT](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#NEXT)
- [PAUSE](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#PAUSE)
- [PLAY](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#PLAY)
- [PREVIOUS](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#PREVIOUS)
- [RECORD](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#RECORD)
- [RED](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#RED)
- [REWIND](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#REWIND)
- [SEARCH](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#SEARCH)
- [SETUP](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#SETUP)
- [SKIP_BACKWARD](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#SKIP_BACKWARD)
- [SKIP_FORWARD](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#SKIP_FORWARD)
- [STOP](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#STOP)
- [SUBTITLE](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#SUBTITLE)
- [VOD](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#VOD)
- [YELLOW](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/ui/Keyboard.html#YELLOW)

Comment (relrelb) [reply core/src/avm2/globals/flash/ui/Keyboard.as]: Nice catch, I removed them all.

**Pull request merged**